drivers/io-expander/demux_spi/demux_spi.c

```diff
@@ -0,0 +1,148 @@
+ *  - The use of this software may or may not infringe the patent rights
+ *    of one or more patent holders.  This license does not release you
+ *    from the requirement that you obtain separate licenses from these
+ *    patent holders to use this software.
+ *  - Use of the software either in source or binary form, must be run
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include "demux_spi.h"
+#include "error.h"
+#include <stdlib.h>
+#include <string.h>
+
+/******************************************************************************/
+/**************************** Types Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief Demux specific SPI platform ops structure
+ */
+const struct spi_platform_ops demux_spi_platform_ops = {
+	.spi_ops_init = demux_spi_init,
+	.spi_ops_remove = demux_spi_remove,
+	.spi_ops_write_and_read = demux_spi_write_and_read
+};
+
+/**
+ * @brief Initialize the SPI demux layer.
+ * @param desc - The SPI descriptor.
+ * @param param - The structure that contains the SPI parameters.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t demux_spi_init(struct spi_desc **desc,
+		       const struct spi_init_param *param)
+{
+	int32_t ret;
+
+	struct spi_desc *descriptor;
+	struct spi_desc *spi_dev_desc;
+	struct spi_init_param *spi_dev_param;
+
+	if (!param)
+		return FAILURE;
+
+	descriptor = (struct spi_desc *)calloc(1, sizeof(*descriptor));
+	if (!descriptor)
+		return FAILURE;
+
+	descriptor->chip_select = param->chip_select;
+	descriptor->max_speed_hz = param->max_speed_hz;
+	descriptor->mode = param->mode;
+
+	spi_dev_param = param->extra;
+
+	ret = spi_init(&spi_dev_desc, spi_dev_param);
+
+	(descriptor->extra) = spi_dev_desc;
```

Comment:
```suggestion
	struct dmux_desc *dmux = calloc(1, sizeof(struct dmux_desc));
	descriptor->extra = dmux;
	
	switch(dmux_param->operation) {
	case INIT_DEMUX:
		dmux->is_dmux_instance = 1;
		//Init next device (either platform device or other dmux)
		spi_init(&dmux->next_dev, dmux_param->param.next_param); //here is platform initialization
		break;
	case ATTACH_SPI_DEVICE:
		dmux->is_dmux_instance = 0;
		descriptor->chip_select = param->chip_select;
		descriptor->max_speed_hz = param->max_speed_hz;
		descriptor->mode = param->mode;

		struct dmux_desc *mux_ref =  dmux_param->param.dmux->extra;
		dmux->next_dev = mux_ref->next_dev;
		break;
	}
```

Reply:
I actually agree with the need for a change like this.
The problem that we have with the current implementation is that we perform the platform's `spi_init` once for each added device. And of course it works as-is because the only difference between various device settings is the chip select number (not the speed, not the mode). Specifically, in projects/fmcjesdadc1, xil_spi_init is called at the initialization of each of the mux driven devices plus the mux itself -> 4 times. The prescaler and SPI mode being identical the code works. But we should avoid this multiple platform `spi_init` call.